data/transactions/logic/assembler_test.go

```diff
@@ -2890,3 +2899,115 @@ int 1
 	`
 	testProg(t, source, AssemblerMaxVersion)
 }
+
+func TestAssembleMatch(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	// fail when target doesn't correspond to existing label
+	source := `
+	pushints 1 1 1
+	match label1 label2
+	label1:
+	`
+	testProg(t, source, AssemblerMaxVersion, NewExpect(3, "reference to undefined label \"label2\""))
+
+	// No labels is pretty degenerate, but ok, I suppose. It's just a no-op
+	testProg(t, `
+int 0
+match
+int 1
+`, AssemblerMaxVersion)
+
+	// confirm arg limit
+	source = `
+	pushints 1 2 1
+	match label1 label2
+	label1:
+	label2:
+	`
+	ops := testProg(t, source, AssemblerMaxVersion)
+	require.Len(t, ops.Program, 12) // ver (1) + pushints (5) + opcode (1) + length (1) + labels (2*2)
+
+	// confirm byte array args are assembled successfully
+	source = `
+	pushbytess "1" "2" "1"
+	match label1 label2
+	label1:
+	label2:
+	`
+	testProg(t, source, AssemblerMaxVersion)
+
+	var labels []string
+	for i := 0; i < 255; i++ {
+		labels = append(labels, fmt.Sprintf("label%d", i))
+	}
+
+	// test that 255 labels is ok
+	source = fmt.Sprintf(`
+	pushint 1
+	match %s
+	%s
+	`, strings.Join(labels, " "), strings.Join(labels, ":\n")+":\n")
+	ops = testProg(t, source, AssemblerMaxVersion)
+	require.Len(t, ops.Program, 515) // ver (1) + pushint (2) + opcode (1) + length (1) + labels (2*255)
+
+	// 256 is too many
+	source = fmt.Sprintf(`
+	pushint 1
+	match %s extra
+	%s
+	`, strings.Join(labels, " "), strings.Join(labels, ":\n")+":\n")
+	testProg(t, source, AssemblerMaxVersion, Expect{3, "match cannot take more than 255 labels"})
+
+	// allow duplicate label reference
+	source = `
+	pushint 1
+	match label1 label1
+	label1:
+	`
+	testProg(t, source, AssemblerMaxVersion)
+}
+
+func TestAssemblePushConsts(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	// allow empty const int list
+	source := `pushints`
+	testProg(t, source, AssemblerMaxVersion)
+
+	// allow empty const bytes list
+	source = `pushbytess`
+	testProg(t, source, AssemblerMaxVersion)
+
+	// basic test
+	source = `pushints 1 2 3`
+	ops := testProg(t, source, AssemblerMaxVersion)
+	require.Len(t, ops.Program, 6) // ver (1) + pushints (5)
+	source = `pushbytess "1" "2" "33"`
+	ops = testProg(t, source, AssemblerMaxVersion)
+	require.Len(t, ops.Program, 10) // ver (1) + pushbytess (9)
+
+	// 256 increases size of encoded length to two bytes
```

Comment:
This probably happens at 128, but I'm fine with this test, it still shows that uvarint encoding is happening. (And I like that we are showing 256 constants can be given.)

Reply:
huh I assumed because if it was a `uvarint` it would go up to `2^8 - 1`. Either way, the test is valid.

Reply:
Basic uvarint encoding: collect a string of bytes such that the high bit of all but the last byte is 1.  Then tack together all the 7 low bits from those bytes.

So as soon as the you get to 2^7, you have to use two bytes.